set_test.go

```diff
@@ -147,3 +235,105 @@ func TestSetIter(t *testing.T) {
 		t.Fatalf("unexpected sum: %d, want %d", sum, sumTo99)
 	}
 }
+
+func testSetMemoryUsage[K IntKey]() func(t *testing.T) {
+	return func(t *testing.T) {
+		s := NewSet[K](16)
+		m := New[K, struct{}](16)
+		szs := int(unsafe.Sizeof(s.data[0]))
+		msz := int(unsafe.Sizeof(m.data[0]))
+		if szs*2 < msz {
+			t.Fatalf("The per-element size of Set[%T] (%d) should be half that of Map[%T] (%d)",
+				s.data[0], szs, m.data[0], msz)
+		}
+		t.Logf("Sizeof(Set[%T]) = %d; Sizeof(Map[%T]) = %d",
+			s.data[0], szs, m.data[0], msz)
+	}
+}
+
+func TestSetMemoryUsage(t *testing.T) {
+	t.Run("int", testSetMemoryUsage[int]())
+	t.Run("int8", testSetMemoryUsage[int8]())
+	t.Run("int16", testSetMemoryUsage[int16]())
+	t.Run("int32", testSetMemoryUsage[int32]())
+	t.Run("int64", testSetMemoryUsage[int64]())
+
+	t.Run("uint", testSetMemoryUsage[uint]())
+	t.Run("uint8", testSetMemoryUsage[uint8]())
+	t.Run("uint16", testSetMemoryUsage[uint16]())
+	t.Run("uint32", testSetMemoryUsage[uint32]())
+	t.Run("uint64", testSetMemoryUsage[uint64]())
+}
+
+func BenchmarkSetAdd10K(b *testing.B) {
+	rr := rand.New(rand.NewSource(12345))
+	seen := make(map[int]struct{}, 10_000)
+	for len(seen) < 10_000 {
+		seen[int(rr.Int31())] = struct{}{}
+	}
+	keys := slices.AppendSeq(make([]int, 0, 10_000), maps.Keys(seen))
+
+	s := NewSet[int](10_000)
+	b.ResetTimer()
+
+	j := 0
+	for i := 0; i < b.N; i++ {
+		s.Add(keys[j])
+		j++
+		if j == 10_000 {
+			j = 0
+			s.Clear()
```

Comment:
This benchmark does not reflect memory usage very well. If we allocate a new Set on every iteration we can see the memory differences with `benchstat`. 

```suggestion
		for range b.N {
			s := NewSet[int](10_000)
			for _, k := range keys {
				s.Add(k)
```
I am itching to use b.Loop() here, but our go.mod is still on 1.23 - so we'll have to wait until we bump to required go version 😅

Reply:
Ah, the purpose of this function is to benchmark only the `Add()` logic (to find hotspots) without any resizing of the  backing slice. The reason I want to avoid resizing here is that then the memory allocator will end up dominating/determining the benchmark results. Additionally, I usually prefer for benchmarks to be per-op (here an individual `Add()`) instead the cumulative time it takes to `Add()` all of the elements (which is what this change suggests), but I don't feel very strongly about this so open to changing it.

As for memory usage the `TestSetMemoryUsage` should cover that / guarantee that it is half that of a `Map`. That said, we could add a benchmark for populating a `Set` that would measure both allocations and memory. So let me know what you think.

Reply:
Ah, ok I see. However -- the allocation strategy is a crucial piece of Add(). If I make a bad change to how memory is allocated, this benchmark would not show it.

If Add() is well implemented, and the size of the Set is big enough, then the allocation overhead should be completely amortized. In particular because we pre-alloc with the known target size. If that is not true, then that would be a regression